File: scala/house/src/main/scala/House.scala
package

object House {
  val Rhymes = """This is the horse and the hound and the horn
    |that belonged to the farmer sowing his corn
    |that kept the rooster that crowed in the morn
    |that woke the priest all shaven and shorn
    |that married the man all tattered and torn
    |that kissed the maiden all forlorn
    |that milked the cow with the crumpled horn
    |that tossed the dog
    |that worried the cat
    |that killed the rat
    |that ate the malt
    |that lay in the house that Jack built."""
      .stripMargin.split('\n').toList ++ List("")

  val RhymesSize = Rhymes.size

  def dropBefore_The(xs: String) = xs.drop(xs.indexOf(" the"))

  def takeRhymes(n: Int) = {
    val rhymes = Rhymes.drop(RhymesSize-1 - n)

    ("This is" ++ dropBefore_The(rhymes.head)) +: rhymes.tail
  }

  lazy val rhyme: String =
    (for(i <- 1 until RhymesSize) yield takeRhymes(i))
      .map(_.mkString("\n"))
      .mkString("\n") + "\n"
}

File: haskell/house/src/House.hs
module House (rhyme) where

import Text.Printf (printf)
import qualified Data.List as L

rhymes :: [(String, String)]
rhymes =
  [ ("_", "the horse and the hound and the horn")
  , ("belonged to", "the farmer sowing his corn")
  , ("kept", "the rooster that crowed in the morn")
  , ("woke", "the priest all shaven and shorn")
  , ("married", "the man all tattered and torn")
  , ("kissed", "the maiden all forlorn")
  , ("milked", "the cow with the crumpled horn")
  , ("tossed", "the dog")
  , ("worried", "the cat")
  , ("killed", "the rat")
  , ("ate", "the malt")
  , ("lay in", "the house that Jack built")
  ]

nRhymes :: Int
nRhymes = length rhymes

rhyme :: String
rhyme = L.intercalate "\n" $ phrase' <$> [0..(nRhymes - 1)]
  where
    it' 0 = "This"
    it' _ = "that"

    does' 0 _ = "is"
    does' _ (x, _) = x

    rhymes' x = drop (nRhymes - x - 1) rhymes

    sing' x xss = let
      xs = xss !! x
      it = it' x
      does = does' x xs
      whatever = snd xs
      in printf "%s %s %s" it does whatever

    phrase' x = let
      xss = rhymes' x
      period y = if y == x then ".\n" else ""
      sing'' z = sing' z xss ++ period z
      in L.intercalate "\n" $ sing'' <$> [0..x]